### ISSUE TITLE
poisson regresson in lightgbm

### ISSUE THREAD

**Issue by spedygiorgio** [Microsoft/LightGBM#233]
Thank you really very much for your amazing package! 
Any plans to introduce:
a. poisson regression as objective?
b. Offset parameter like in poisson regression?

This would permit to handle variable exposure observations in count regressions

**Comment by henry0312** [comment]
Please give me the definition of poisson regression as objective.
When I have free time, I'll add if possible.

**Comment by spedygiorgio** [comment]
Hi thank you very much for your reply. I think that regarding point one you can have a look to xgboost "poisson regression" as objective [xgboost](https://github.com/dmlc/xgboost/blob/master/doc/parameter.md) (see also the max_delta_step_parameter). Regarding the offset parameter it depends: 
a. the gbm r package has an offset parameter explicity,
b. the approach used in xgboost to obtain the same result (well I got the same result in a similar applicaton) can be found in this SO question (http://stats.stackexchange.com/questions/223939/poisson-xgboost-with-exposure)[stackoverflow]

**Comment by henry0312** [comment]
Is the loss definition this, right?
(from https://en.wikipedia.org/wiki/Poisson_regression)

![image](https://cloud.githubusercontent.com/assets/433434/22321980/2fee7e1a-e3dc-11e6-8a70-3319bc97ba71.png)

**Comment by spedygiorgio** [comment]
I think so but I am not sure.

**Comment by Laurae2** [comment]
@henry0312 Using that formula provides no direct solution. You need to go for negative log-likelihood so gradient descent can be applied.

Therefore, you would need an exponential and you end up with:

* Gradient: `exp(pred) - label`
* Hessian (gradient derivative by pred): `exp(pred)`

Due to `exp(pred)` as the hessian, stability issues can occur quickly, in both sides (low values or gigantic values). Therefore, for satisfying convergence and stability, you would use an offset to the hessian. xgboost uses 0.7 for instance (`exp(pred) + 0.7`, where `max_delta_step = 0.7`), but there could be better / worse values.

**Comment by spedygiorgio** [comment]
Thanks, BTW are also you planning to support offset?

Il lun 30 gen 2017, 17:29 Guolin Ke <notifications@github.com> ha scritto:

**Comment by Laurae2** [comment]
@spedygiorgio It is already supported. You must use the bias (initial score in LightGBM).

Do remember in the case of Poisson regression + non-linear model to modify accordingly either the label or the bias.

Let's suppose you have claims and account count, either is right:

* `y = claims/accounts`, `x = ...`
* `y = claims`, `x = ...`, `base score = log(accounts)`
